crux-mir-comp/src/Mir/Compositional/Override.hs

```diff
@@ -0,0 +1,628 @@
+                let shp = tyToShapeEq col ty (ptr ^. mpType)
+                matchArg sym eval (ms ^. MS.csPreState . MS.csAllocs) shp rv sv
+
+
+        -- Validity checks
+
+        -- All pre-state and post-state fresh vars must be bound.
+        termSub <- use MS.termSub
+        let allFresh = ms ^. MS.csPreState . MS.csFreshVars ++
+                ms ^. MS.csPostState . MS.csFreshVars
+        forM_ allFresh $ \tec -> do
+            let var = SAW.ecVarIndex $ SAW.tecExt tec
+            when (not $ Map.member var termSub) $ do
+                error $ "argument matching failed to produce a binding for " ++
+                    show (MS.ppTypedExtCns tec)
+
+        -- All pre-state allocs must be bound.
+        allocSub <- use MS.setupValueSub
+        forM_ (Map.toList $ ms ^. MS.csPreState . MS.csAllocs) $ \(alloc, info) -> do
+            when (not $ Map.member alloc allocSub) $ do
+                error $ "argument matching failed to produce a binding for " ++
+                    show alloc ++ " (info: " ++ show info ++ ")"
+
+        -- All references in `allocSub` must point to disjoint memory regions.
+        liftIO $ checkDisjoint sym (Map.toList allocSub)
+
+        -- TODO: see if we need any other assertions from LLVM OverrideMatcher
+
+
+        -- Handle preconditions and postconditions.  
+
+        -- Convert preconditions to `osAsserts`
+        forM_ (ms ^. MS.csPreState . MS.csConditions) $ \cond -> do
+            term <- condTerm sc cond
+            w4VarMap <- liftIO $ readIORef w4VarMapRef
+            pred <- liftIO $ termToPred sym sc w4VarMap term
+            MS.addAssert pred $
+                SimError loc (AssertFailureSimError (show $ W4.printSymExpr pred) "")
+
+        -- Convert postconditions to `osAssumes`
+        forM_ (ms ^. MS.csPostState . MS.csConditions) $ \cond -> do
+            term <- condTerm sc cond
+            w4VarMap <- liftIO $ readIORef w4VarMapRef
+            pred <- liftIO $ termToPred sym sc w4VarMap term
+            MS.addAssume pred
+
+    ((), os) <- case result of
+        Left err -> error $ show err
+        Right x -> return x
+
+    forM_ (os ^. MS.osAsserts) $ \lp ->
+        liftIO $ addAssertion sym lp
+
+    forM_ (os ^. MS.osAssumes) $ \p ->
+        liftIO $ addAssumption sym $ W4.LabeledPred p $
+            AssumptionReason loc "methodspec postcondition"
+
+    let preAllocMap = os ^. MS.setupValueSub
+
+    let postAllocDefs = filter (\(k,_v) -> not $ Map.member k preAllocMap) $
+            Map.toList $ ms ^. MS.csPostState . MS.csAllocs
+    postAllocMap <- liftM Map.fromList $ forM postAllocDefs $ \(alloc, Some allocSpec) -> do
+        ref <- newMirRefSim (allocSpec ^. maType)
+        return (alloc, Some $ MirPointer (allocSpec ^. maType) ref)
+    let allocMap = preAllocMap <> postAllocMap
+
+    -- Handle return value and post-state PointsTos
+
+    -- TODO: clobber all writable memory that's accessible in the pre state
```

Comment:
How easily could this be implemented now? Or should we leave it until later?

Reply:
I actually did it already, but forgot to remove the comment.  The mechanism is a little bit indirect: for each accessible location, `add_arg` not only adds a pre-state `PointsTo` (representing a constraint on the contents of the initial memory), but also adds a post-state `PointsTo` that causes the value to be overwritten with a fresh symbolic variable.